### PR TITLE
Add controller to disable CPMS on ARO clusters

### PR DESCRIPTION
**Pull request by tsatam**
### Which issue this PR addresses:

Fixes [ARO-8836](https://issues.redhat.com/browse/ARO-8836)

### What this PR does / why we need it:

- [x] Updates openshift/api dependency to release-4.12 (to bring in CPMS struct)
- [x] Adds a new operatorflag to enable/disable CPMSDeactivator controller
- [x] Adds new CPMSDeactivator controller
- [x] Updates SRE RBAC to allow managing CPMS resource

### Test plan for issue:

- [x] Local 4.12 cluster installs w/ new operator and observes no change in behavior
- [x] Local 4.13+ cluster installs w/ new operator and disables CPMS
- [x] Local <=4.11 cluster (before the introduction of the CPMS resource) can be updated to new operator and observes no change in behavior

### Is there any documentation that needs to be updated for this PR?

No 

### How do you know this will function as expected in production? 

- Tests performed above 

### Additional notes
We should add an E2E test that ensures CPMS is disabled. I did not do it within this PR as it would require another dependency bump+revendor for openshift/client-go.

### CODE REVIEW

**Review by tsatam — COMMENTED** [review]
Feedback from sync PR review session

File: go.mod
Comment (hlipsig) [inline go.mod:277]: Non blocker, but why did this need to be bumped?

File: pkg/operator/flags.go
Comment (hlipsig) [inline pkg/operator/flags.go:13]: Does this turn machinesets off or on? Name is confusing can we make it clearer.

**Review by hlipsig — CHANGES_REQUESTED** [review]
Let's talk about the controller name before merging.

File: go.mod
Comment (tsatam) [reply go.mod:277]: I needed to bring in openshift/api release-4.12 or newer to have access to the ControlPlaneMachineSet struct.

Comment (hlipsig) [reply go.mod:277]: Thanks for the clarification. Not that keeping dependencies current is at all bad, but I didn't see the exact cause immediately in my review so I was just curious.

File: pkg/operator/flags.go
Comment (tsatam) [reply pkg/operator/flags.go:13]: Happy for better suggestions for the name. 

Enabling the controller disables the machineset, which is inherently confusing.

**Review by hlipsig — APPROVED** [review]

File: cmd/aro/operator.go
Comment (s-fairchild) [inline cmd/aro/operator.go:242]: Not really a big deal, but ultimately we want `err` as a string, may as well do it directly IMO.  

```suggestion
				return fmt.Errorf("unable to create controller %s: %s", cpms.ControllerName, err.String())
```

File: cmd/aro/operator.go
Comment (s-fairchild) [inline cmd/aro/operator.go:245]:
```suggestion
			log.Infof("Cluster does not support machinev1 API, will not set up CPMS controller: %s", err.String())
```

File: pkg/operator/controllers/cpms/cpms_controller.go
Comment (s-fairchild) [inline pkg/operator/controllers/cpms/cpms_controller.go:81]: Should we validate cpms is in an inactive state after deleting it?
I see we do that check beforehand, the next time the controller reconciles it will do the check.  
I'm just concerned about possible scenarios where cpms is getting recreated by something/someone, and our controller will just keep deleting it without raising a red flag.

Idea, maybe we can keep a tally somewhere of how many times this controller has deleted cpms, and if it's more than X(1?) generate an error.

Per the docs:
> Verification
Check the control plane machine set custom resource state. A result of Inactive indicates that the removal and replacement process is successful. A ControlPlaneMachineSet CR exists but is not activated.

**Review by s-fairchild — COMMENTED** [review]
Great work here, just a few minor suggestions, and thoughts on verifying directly after deleting cpms.

File: pkg/operator/controllers/cpms/cpms_controller.go
Comment (tsatam) [reply pkg/operator/controllers/cpms/cpms_controller.go:81]: In the interest of keeping this controller in-line with controller recommendations (do one thing per reconcile loop), and the fact that this would introduce a dependency on this controller waiting for CPMSO to roll out a new Inactive CPMS, I'd rather not introduce a "wait" into the controller itself. 

The controller will reconcile against the newly-created inactive CPMS, and log that. IMO if we want to gather any data/metrics around CPMS being active/inactive, we should do so in a monitor rather than in this operator controller.

File: cmd/aro/operator.go
Comment (tsatam) [reply cmd/aro/operator.go:242]: This is an interesting suggestion, but it's not in line with the conventions we have been following throughout the codebase (e.g. https://github.com/Azure/ARO-RP/blob/e711e610a0ad352ac11d0e72328170d759638242/pkg/operator/controllers/workaround/workaround_controller.go#L72). I think if we want to do this, it should be done in a dedicated "style-fixup" PR.

File: pkg/operator/controllers/cpms/cpms_controller.go
Comment (tsatam) [inline pkg/operator/controllers/cpms/cpms_controller.go:27]: rename this

File: cmd/aro/operator.go
Comment (SudoBrendan) [reply cmd/aro/operator.go:242]: a lot of our code uses `%v`, but it is rare that `%v` or `%s` are best here, we should use `%w` unless we care for some reason that the Error type we return is distinct from `err` itself: https://stackoverflow.com/questions/61283248/format-errors-in-go-s-v-or-w

`%s` vs `%v` is also discussed in that thread - the biggest concern is if `err` is nil, which shouldn't happen...?

File: pkg/operator/controllers/cpms/cpms_controller.go
Comment (SudoBrendan) [inline pkg/operator/controllers/cpms/cpms_controller.go:46]: I think this comment is the perfect place in all this code to _not_ use `CPMS` and explicity state it as `Control Plane Machine Set` so we codify the meaning we are after very clearly without having to read the whole function and flow of the operator/controller.

File: pkg/operator/flags.go
Comment (SudoBrendan) [inline pkg/operator/flags.go:13]: nit: this may become a confusing name in a world where we _do_ have OCP CPMS enabled en-masse for the fleet. We should probably use a name like `CPMSDisablerEnabled`

File: pkg/operator/flags.go
Comment (SudoBrendan) [inline pkg/operator/flags.go:51]: nit: same naming issues as above.

**Review by SudoBrendan — CHANGES_REQUESTED** [review]
Overall LGTM - ping me in slack when the new names are here it might clarify my comments up too.

File: pkg/operator/controllers/cpms/cpms_controller.go
Comment (SudoBrendan) [inline pkg/operator/controllers/cpms/cpms_controller.go:27]: nit: similar to my comments below, I think `CPMSDisablerController` is a better name, more future-proof to when we use CPMS across the fleet (if that happens)

File: pkg/operator/flags.go
Comment (tsatam) [reply pkg/operator/flags.go:13]: I had that before and we decided to move away from that name due to it being confusing the other way round: https://github.com/Azure/ARO-RP/pull/3660/commits/c44d92065a0917ba667ec7f830453e746f226738

**Review by SudoBrendan — CHANGES_REQUESTED** [review]

**Review by SudoBrendan — APPROVED** [review]
Overall LGTM - feel free to address and/or close any remaining comments.

File: pkg/operator/controllers/cpms/cpms_controller.go
Comment (tsatam) [reply pkg/operator/controllers/cpms/cpms_controller.go:27]: We plan to use this controller to operationalize CPMS once we're ready for it.

File: pkg/operator/controllers/cpms/cpms_controller.go
Comment (SudoBrendan) [inline pkg/operator/controllers/cpms/cpms_controller.go:56]: nit: might be worth a comment here stating that this controller is unique in that `enabled: true` means that CPMS sticks around, and `enabled: false` means that we will deactivate it. I think it is highly clear once you see it - but it is also distinct.

Comment (SudoBrendan) [reply pkg/operator/controllers/cpms/cpms_controller.go:56]: alternatively, I see the comment in the header of this function, doing some `**false**` formatting or something may draw attention to the uniqueness.

File: cmd/aro/operator.go
Comment (s-fairchild) [reply cmd/aro/operator.go:242]: Thanks for the link. I forgot there was an error formatter, that would be ideal.

Comment (tsatam) [reply cmd/aro/operator.go:242]: The entirety of cmd/aro/operator.go is a prime candidate for some `%w` replacement on error wrapping - I think this should be done in a follow-up PR across _all_ of our controller instantiation.

**Review by kimorris27 — APPROVED** [review]
LGTM, but it would be nice to see an E2E test in a follow-up PR.

**Review by hlipsig — APPROVED** [review]
Great discussions. LGTM as is and the improvements we can card up for the future.

**Pull request merged**
